staging/src/k8s.io/client-go/tools/cache/shared_informer_health.go

```diff
@@ -0,0 +1,214 @@
+		}
+	}
+}
+
+func (lw *listWatchHealth) health() (int, bool) {
+	lw.lock.Lock()
+	defer lw.lock.Unlock()
+	if lw.startingTime.IsZero() {
+		return 0, true
+	}
+	now := time.Now()
+	elapsed := now.Sub(lw.startingTime)
+	if elapsed > lw.minTimeToFail {
+		return lw.consistentErrors, false
+	}
+	return 0, true
+}
+
+func newDefaultListWatchLockFreeHealth(errorReceiverCh <-chan error) *listWatchLockFreeHealth {
+	lwlf := &listWatchLockFreeHealth{
+		minTimeToFail:       time.Duration(5 * time.Second),
+		minConsistentErrors: 10,
+		consistentErrorsCh:  errorReceiverCh,
+		errCircularBuffer:   atomic.Value{},
+	}
+
+	lwlf.errCircularBuffer.Store(newErrBuffer(5))
+	return lwlf
+}
+
+type listWatchLockFreeHealth struct {
+	startingTime int64
+
+	// minTimeToFail minimum time duration after which we start to fail
+	minTimeToFail time.Duration
+	// minConsistentError minimum number of errors before we start to fail
+	minConsistentErrors int32
+	// consistentErrors holds the number of errors seen so far
+	consistentErrors int32
+
+	// consistentErrorsCh channel on which we receive List&Watch related errors
+	// it is meant to represent a gauge, that is a number that can go up and down
+	consistentErrorsCh <-chan error
+
+	errCircularBuffer atomic.Value
+}
+
+func (lwlf *listWatchLockFreeHealth) run(stopCh <-chan struct{}) {
+	//TODO: use wait.Until(...)
+	go lwlf.runInternal(stopCh)
+}
+
+func (lwlf *listWatchLockFreeHealth) runInternal(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case errRcv, ok := <-lwlf.consistentErrorsCh:
+			if !ok {
+				break
+			}
+			seenErrors := atomic.LoadInt32(&lwlf.consistentErrors)
+			if errRcv == nil {
+				seenErrors = 0
+			} else {
+				seenErrors = seenErrors + 1
+				errBuf := lwlf.errCircularBuffer.Load().(errBuffer)
+				errBufCpy := errBuf.copy()
+				errBufCpy.add(errRcv)
+				lwlf.errCircularBuffer.Store(errBufCpy)
+			}
+			localStartingTime := atomic.LoadInt64(&lwlf.startingTime)
+			atomic.StoreInt32(&lwlf.consistentErrors, seenErrors)
+			if seenErrors >= lwlf.minConsistentErrors && localStartingTime <= 0 {
+				atomic.StoreInt64(&lwlf.startingTime, time.Now().UnixNano())
+			} else if seenErrors < lwlf.minConsistentErrors && localStartingTime > 0 {
+				atomic.StoreInt64(&lwlf.startingTime, 0)
+			}
+		}
+	}
+}
+
+func (lwlf *listWatchLockFreeHealth) health() (int, []error, bool) {
+	localStartingTimeUnix := atomic.LoadInt64(&lwlf.startingTime)
+	localConsistentErrors := atomic.LoadInt32(&lwlf.consistentErrors)
+	errBuf := lwlf.errCircularBuffer.Load().(errBuffer)
+	if localStartingTimeUnix <= 0 {
+		return 0, nil, true
+	}
+	if localConsistentErrors < lwlf.minConsistentErrors {
+		return 0, nil, true
+	}
+	now := time.Now()
+	startingTime := time.Unix(0, localStartingTimeUnix)
+	elapsed := now.Sub(startingTime)
+	if elapsed > lwlf.minTimeToFail {
+		return int(localConsistentErrors), errBuf.get(), false
+	}
+	return 0, nil, true
+}
+
+type errBuffer struct {
```

Comment:
I'm surprised there isn't something off the shelf.  You'll need a fair number of unit tests for this.

Reply:
I'm pretty sure we could find something on the internet but nothing in `kubernetes` repo. Do you have any specific tests in your mind ?